src/contexts/NetworkContext.js

```diff
@@ -0,0 +1,61 @@
+import React, { createContext, useState, useEffect, useContext } from 'react';
+import storage from '../services/storage';
+
+const NetworkConext = createContext();
+
+/**
+ * Custom hook to subscribe to network status.
+ *
+ * Value is retrieved from local storage initially and updated by service worker firing events
+ *
+ * Can be used by any functional component under the NetworkContextProvider in App.js
+ *
+ * @returns {boolean} true if connected to the network, false otherwise.
+ */
+export const useNetworkStatus = () => {
+  const context = useContext(NetworkConext);
+  if (context === undefined) {
+    throw new Error(`useNetworkStatus must be called within NetworkContextProvider`);
+  }
+
+  return context;
+};
+
+const NetworkContextProvider = ({ children }) => {
+  const [isOnline, setIsOnline] = useState('online');
+
+  useEffect(() => {
+    // Retrieve network status from local storage or default to online
+    try {
+      const networkStatus = storage.get('network-status');
+      setIsOnline(networkStatus === 'online');
+    } catch (error) {
+      console.error(error);
+      setIsOnline(true);
+    }
+
+    /* Used to re-render the page when the network connection changes.
+     * The origin of the message is checked to prevent cross-site scripting attacks
+     */
+    const updateNetworkStatus = (event) => {
+      setIsOnline((prevStatus) => {
+        if (
+          event.origin === window.location.origin &&
+          (event.data === 'online' || event.data === 'offline')
+        ) {
+          return event.data === 'online';
+        } else {
+          return prevStatus;
+        }
+      });
+    };
+
+    window.addEventListener('message', updateNetworkStatus);
+
+    return () => window.removeEventListener('message', updateNetworkStatus);
```

Comment:
With my meager understanding of event listeners, I cannot follow what is happening here. If you have the time to explain this, I would appreciate it.

Reply:
This is essentially the same as the existing network-status code (which will be refactored out of the ~30 components that repeat it in the near future hopefully). It is basically attaching an event listener callback (`updateNetworkStatus` in this case) to the `window`, aka the browser tab/window that 360 is running in. 

This event listener callback will be called whenever an event is sent to the browser that has the 'type' (which is a string describing the kind of event) of `message`. These `message` events are sent either by our serviceWorker or the browser itself (although any other JavaScript code could send such an event, which is why we have to make sure in the event listener callback that we only pay attention to events originating from within our site). The callback is passed the `event` that triggered the eventListener, which in this case should include data about the status of the network connection (i.e. 'online' or 'offline'). 

Line 55 is telling the `useEffect` hook to call that arrow function whenever the NetworkContextProvider component is unmounted. Since the provider is mounted at the root of our app, this is unlikely to ever happen, but is still best practice for eventListeners. The arrow function simply removes the event listener from the `window`, allowing the browser to forget about it. If you don't remove the event listener,  you're creating a memory leak where the browser maintains an ever-expanding list of event listeners that were used by components which no longer exist and are therefore useless but are taking up memory and slowing down the browser.